openfl-tutorials/interactive_api/MXNet_landmarks/envoy/landmark_shard_descriptor.py

```diff
@@ -0,0 +1,172 @@
+                 **kwargs):
+        """Initialize LandmarkShardDescriptor."""
+        super().__init__()
+        # Settings for sharding the dataset
+        self.rank, self.worldsize = map(int, rank_worldsize.split(','))
+
+        self.data_folder = Path.cwd() / data_folder
+        self.download_data()
+
+        # Calculating data and target shapes
+        ds = self.get_dataset()
+        sample, target = ds[0]
+        self._sample_shape = [str(dim) for dim in sample.shape]
+        self._target_shape = str(len(target.shape))
+
+        assert self._target_shape == '1', 'Target shape Error'
+
+    def process_data(self, name_csv_file):
+        """Process data from csv to numpy format and save it in the same folder."""
+        data_df = pd.read_csv(self.data_folder / name_csv_file)
+        data_df.fillna(method='ffill', inplace=True)
+        keypoints = data_df.drop('Image', axis=1)
+        cur_folder = str(self.data_folder.relative_to(Path.cwd())) + '/'
+
+        for i in range(data_df.shape[0]):
+            img = data_df['Image'][i].split(' ')
+            img = np.array(['0' if x == '' else x for x in img], dtype='float32').reshape(96, 96)
+            np.save(cur_folder + 'img_' + str(i) + '.npy', img)
+            y = np.array(keypoints.iloc[i, :], dtype='float32')
+            np.save(cur_folder + 'keypoints_' + str(i) + '.npy', y)
+
+    def download_data(self):
+        """Download dataset from Kaggle."""
+        if not os.path.exists(self.data_folder):
+            os.mkdir(self.data_folder)
+
+        if not self.is_dataset_complete():
+            logger.info('Your dataset is absent or damaged. Downloading ... ')
+            api = KaggleApi()
+            api.authenticate()
+
+            if os.path.exists('data/train'):
+                shutil.rmtree('data/train')
+
+            api.competition_download_file(
+                'facial-keypoints-detection',
+                'training.zip', path=self.data_folder
+            )
+
+            with ZipFile(self.data_folder / 'training.zip', 'r') as zipobj:
+                zipobj.extractall(self.data_folder)
+
+            os.remove(self.data_folder / 'training.zip')
+
+            self.process_data('training.csv')
+            os.remove(self.data_folder / 'training.csv')
+            self.save_all_md5()
+
+    def get_dataset(self, dataset_type='train'):
+        """Return a shard dataset by type."""
+        return LandmarkShardDataset(
+            dataset_dir=self.data_folder,
+            rank=self.rank,
+            worldsize=self.worldsize
+        )
+
+    def calc_all_md5(self):
+        """Calculate hash of all dataset."""
+        md5_dict = {}
+        for root, _, files in os.walk(self.data_folder):
+            for file in files:
+                if file == 'dataset.json':
+                    continue
+                md5_calc = md5()
+                rel_dir = os.path.relpath(root, self.data_folder)
+                rel_file = os.path.join(rel_dir, file)
```

Comment:
Could you please use `pathlib` where it possible

Reply:
Ok, but why? We shouldn't use the 'os' library at all?